lecture16/library/pageobjects/page_objects.py

```diff
@@ -0,0 +1,112 @@
+import time
+from typing import List
+
+from custom_selectors.my_selectors import *
+from models.product import Product
+from models.user import User
+from selenium.common.exceptions import NoSuchElementException
+from selenium.webdriver.chrome.webdriver import WebDriver
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+
+
+class MainPage:
+    url = "https://www.demoblaze.com/"
+
+    def __init__(self, driver: WebDriver):
+        self.driver = driver
+
+    def click_product_by_title(self, title):
+        product_titles = self.driver.find_elements(*MainPageSelectors.PRODUCTS_TITLES)
+        matching_titles = [pt for pt in product_titles if pt.text == title]
+        if matching_titles:
+            matching_titles[0].click()
+        else:
+            raise NoSuchElementException(f'Element with title {title} was not found')
+
+    def click_monitors_category(self):
+        WebDriverWait(self.driver, 10).until(
+            EC.visibility_of_element_located(MainPageSelectors.MONITORS_CATEGORY)).click()
+
+    def get_product_with_highest_price(self) -> Product:
+        # Necessary waiting, because there is nothing on the page to bind to when switching between categories
+        time.sleep(3)

```

Comment:
Wait Until Keyword Succeeds is a built-in RF keyword that can be used instead of sleep
In this case there is also a problem of double unnecessary waiting when calling get_highest_price_product_price after get_highest_price_product_title
There is a way to set cached property https://pypi.org/project/cached-property/